mitodl/open-discussions
# refactor course search state to be driven off of URL

#### Pre-Flight checklist


#### What are the relevant tickets?

closes #2333

#### What's this PR do?

this PR basically refactors how we store the search state for the course search page so that the params in the URL are the sole source of truth for the filters and so on.

right now in master these things are sort of stored in two places - in the URL bar and in the component state for the `CourseSearchPage` component. We have some sort of weird logic for syncing them, and for getting param values from the url bar on startup.

This makes it so that a few things (like incremental vs full-refresh searching) are driven off of values kept in component state while filters and all that stuff are driven purely off of the URL bar. I think this makes it simpler - hopefully you agree!

#### How should this be manually tested?

There should be no regressions in user experience or behavior on the course search page. things should work!

And definitely read through all the code to make sure it makes sense. It's a fairly involved refactor - I think I dotted all the 'i's and crossed the 't's but I easily could have missed a few :smiley:
